include/deal.II/numerics/vector_tools_evaluate.h

```diff
@@ -0,0 +1,252 @@
+// ---------------------------------------------------------------------
+//
+// Copyright (C) 2021 by the deal.II authors
+//
+// This file is part of the deal.II library.
+//
+// The deal.II library is free software; you can use it, redistribute
+// it, and/or modify it under the terms of the GNU Lesser General
+// Public License as published by the Free Software Foundation; either
+// version 2.1 of the License, or (at your option) any later version.
+// The full text of the license can be found in the file LICENSE.md at
+// the top level directory of deal.II.
+//
+// ---------------------------------------------------------------------
+
+
+#ifndef dealii_vector_tools_evaluation_h
+#define dealii_vector_tools_evaluation_h
+
+#include <deal.II/base/config.h>
+
+#include <deal.II/base/mpi_remote_point_evaluation.h>
+
+#include <deal.II/dofs/dof_handler.h>
+
+#include <deal.II/fe/fe_point_evaluation.h>
+
+#include <map>
+
+DEAL_II_NAMESPACE_OPEN
+
+namespace VectorTools
+{
+  namespace EvaluationFlags
+  {
+    enum EvaluationFlags
+    {
+      avg    = 0,
+      max    = 1,
+      min    = 2,
+      insert = 3
+    };
+  }
+
+  /**
+   * Given a (distributed) solution vector @p vector, evaluate the values at
+   * the (arbitrary and even remote) points specified by @p evaluation_points.
+   */
+  template <int n_components, int dim, int spacedim, typename VectorType>
+  std::vector<typename FEPointEvaluation<n_components, dim>::value_type>
+  evaluate_at_points(
```

Comment:
> In terms of technicalities, I think we can return any information FEPointEvaluation can compute, as long as we can pack the data appropriately.

Yes that is right. I would suggest once other quantities are needed, we move the content of this function to an internal (templated) function, which is called by `evaluate_values_at_points()`, `evaluate_gradients_at_points()`, ...

> I think we should also think about the case where a user requests the gradient of a solution field (or the Hessian). How would we structure the case of various return kinds? Set up some additional data structure that is filled upon demand?

The bigger question is how we can deal with derived quantities (e.g. in the case of SIM I would like to communicate already the force computed with `JxW` so that only has to be tested after communication) and with composites (e.g., you might want to pack multiple quantities (one might be a scalar, the other one vectorial) together so that they are sent in one go; e.g. BDF-2 multiple timesteps after AMR step).

The best solution I was able to come up (in adaflo) for this till now is use `RemovePointEvaluation` directly. This class can work with arbitrary types so that I can also pass in vectors of tuples. What is your opinion?